### PR TITLE
vtgate load data: support load data

### PR DESCRIPTION
**Pull request by cjsheng**
The target of Implementing the "load data local infile"  on "vtgate" is to solve the problem of heterogeneous systems importing data to vitess (eg: hive->vitess).

I implemented the load data  in  vtgate  and converted the data to "insert into Ignore".

### CODE REVIEW

File: go/mysql/query.go [hunk 1]
```diff
@@ -484,6 +484,37 @@ func (c *Conn) sendColumnCount(count uint64) error {
...
 
+// calcPacketSize calculate size of request file package
+func calcPacketSize(fileName []byte) int {
+	if fileName == nil {
+		return 1
+	}
+	return 1 + len(fileName)
+}
+
+//WriteRequestFilePacket  write the package of request file
```
Comment (sougou) [inline go/mysql/query.go]: Space between `/` and `W`. Same for other comments in file.

File: go/vt/sqlparser/ast.go [hunk 2]
```diff
@@ -536,6 +537,61 @@ type InsertRows interface {
...
+	FieldsInfo *FieldsClause
+	LinesInfo  *LinesClause
+}
+
+// Format formats the node.
+func (node *LoadDataStmt) Format(buf *TrackedBuffer) {
+}
+
+// walkSubtree walks the nodes of the subtree.
+func (node *LoadDataStmt) walkSubtree(visit Visit) error {
```
Comment (sougou) [inline go/vt/sqlparser/ast.go]: walk needs to walk the subnodes.

File: go/vt/sqlparser/escape.go [hunk 3]
```diff
@@ -0,0 +1,22 @@
+package sqlparser
+
+// EscapeChar handle the mysql  escape characters
+func EscapeChar(c byte) byte {
```
Comment (sougou) [inline go/vt/sqlparser/escape.go]: This should instead reuse `SQLEncodeMap` in go/sqltypes/value.go.

**Comment by sougou** [comment]
For passing through, this is what I had in mind (working backwards):
* The `load data infile` command will be passed through to all vttablets via streaming protocol.
* vttablet will send it to mysql, which will send a request packet back to vttablet.
* vttablet will send an 'ack' back which will be collected by vtgate from all vttablets.
* Once ack is collected, vtgate will send the request packet command to client.
* Client will start sending rows.
* vtgate analyzes each row, groups and routes them to the appropriate vttablet.
* Client sends 'done', which is then transmitted as 'done' to all vttablets.
* vttablets send 'done' to all mysqls.

It's complicated, but I don't think we need temp files if we used the above process. Then we can load data into vitess at super-high-speed.

File: go/vt/vtgate/load_data.go [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+		line = line[1 : len(line)-1]
+		sep = make([]byte, 0, len(l.FieldsInfo.Terminated)+2)
+		sep = append(sep, l.FieldsInfo.Enclosed)
+		sep = append(sep, l.FieldsInfo.Terminated...)
+		sep = append(sep, l.FieldsInfo.Enclosed)
+	} else {
+		sep = []byte(l.FieldsInfo.Terminated)
+	}
+	rawCols := bytes.Split(line, sep)
+	cols := escapeCols(rawCols)
```
[the remark this answers is not in the document]
Comment (Johnny-Three) [reply go/vt/vtgate/load_data.go]: what's the way to make it more safer ?

**Comment by morgo** [comment]
I am going to close this PR due to staleness. The feature `LOAD DATA INFILE` is being tracked in https://github.com/vitessio/vitess/issues/2976

There are known security issues regarding this feature. In discussion with other maintainers, we propose implementation as a utility that batch loads text files, converting to standard SQL for execution by MySQL. The performance is usually similar to `LOAD DATA`, since MySQL does actually do that much to optimize this statement.

**Pull request closed without merging**